internal/api/callbacks.go

```diff
@@ -310,6 +316,55 @@ func cNext(ref C.iterator_t, gasMeter *C.gas_meter_t, usedGas *cu64, key *C.Unma
 	return C.GoError_None
 }
 
+//export cNextKey
+func cNextKey(ref C.iterator_t, gasMeter *C.gas_meter_t, usedGas *cu64, key *C.UnmanagedVector, errOut *C.UnmanagedVector) (ret C.GoError) {
+	return nextPart(ref, gasMeter, usedGas, key, errOut, func(iter types.Iterator) []byte { return iter.Key() })
+}
+
+//export cNextValue
+func cNextValue(ref C.iterator_t, gasMeter *C.gas_meter_t, usedGas *cu64, value *C.UnmanagedVector, errOut *C.UnmanagedVector) (ret C.GoError) {
+	return nextPart(ref, gasMeter, usedGas, value, errOut, func(iter types.Iterator) []byte { return iter.Value() })
+}
+
+// nextPart is a helper function that contains the shared code for key- and value-only iteration.
+func nextPart(ref C.iterator_t, gasMeter *C.gas_meter_t, usedGas *cu64, output *C.UnmanagedVector, errOut *C.UnmanagedVector, valFn func(types.Iterator) []byte) (ret C.GoError) {
+	// typical usage of iterator
+	// 	for ; itr.Valid(); itr.Next() {
+	// 		k, v := itr.Key(); itr.Value()
+	// 		...
+	// 	}
+
+	defer recoverPanic(&ret)
+	if ref.call_id == 0 || gasMeter == nil || usedGas == nil || output == nil || errOut == nil {
+		// we received an invalid pointer
+		return C.GoError_BadArgument
+	}
+	if !(*output).is_none || !(*errOut).is_none {
+		panic("Got a non-none UnmanagedVector we're about to override. This is a bug because someone has to drop the old one.")
+	}
+
+	gm := *(*types.GasMeter)(unsafe.Pointer(gasMeter))
+	iter := retrieveIterator(uint64(ref.call_id), uint64(ref.iterator_index))
+	if iter == nil {
+		panic("Unable to retrieve iterator.")
+	}
+	if !iter.Valid() {
+		// end of iterator, return as no-op, nil key is considered end
+		return C.GoError_None
+	}
+
+	gasBefore := gm.GasConsumed()
+	// call Next at the end, upon creation we have first data loaded
+	out := valFn(iter)
+	// check iter.Error() ????
+	iter.Next()
+	gasAfter := gm.GasConsumed()
+	*usedGas = (cu64)(gasAfter - gasBefore)
```

Comment:
Is this another case where Cosmos SDK gas is not converted to CosmWasm gas but should be? Looks suspecious to me. Same for the regular next case above.

Reply:
Only if the `GasMeter` that gets passed by wasmd is not a `MultipliedGasMeter`. I just checked and as far as I can see, all of them are multiplied. But I feel like using a different interface here would have been safer (e.g. with a `WasmGasConsumed` function) to make it impossible to pass an sdk `GasMeter`.